File: backend/app/reporter.py
import json
import logging
import os
from typing import Any

import httpx

logger = logging.getLogger(__name__)


async def generate_report(
    variant: dict[str, Any],
    annotations: dict[str, Any],
    similar_variants: list[dict[str, Any]],
    structures: dict[str, Any],
) -> dict[str, Any]:
    """Generate final report content.

    Contributor task:
    - Use Claude with strict instructions to only use structured evidence.
    - Return markdown plus structured JSON for frontend rendering.
    """
    evidence = _structured_evidence(variant, annotations, similar_variants, structures)
    ai_markdown = await _claude_report(evidence) or await _groq_report(evidence)
    markdown = ai_markdown or _fallback_markdown(evidence)

    return {
        "markdown": markdown,
        "json": {
            "status": "complete",
            "variant": evidence["variant"],
            "classification_summary": evidence["classification_summary"],
            "wild_type": evidence["wild_type"],
            "unknown_variant": evidence["unknown_variant"],
            "similar_variants": evidence["similar_variants"],
            "structures": evidence["structures"],
            "warnings": evidence["warnings"],
            "disclaimer": (
                "Research support only. This is not medical advice, diagnosis, or treatment guidance."
            ),
        },
    }


def _structured_evidence(
    variant: dict[str, Any],
    annotations: dict[str, Any],
    similar_variants: list[dict[str, Any]],
    structures: dict[str, Any],
) -> dict[str, Any]:
    parsed_variant = annotations.get("variant") or variant
    features = annotations.get("features") or {}
    alpha = annotations.get("alpha_missense") or {}
    clinvar_records = annotations.get("clinvar", {}).get("records") or []
    warnings = []
    warnings.extend(annotations.get("warnings") or [])
    warnings.extend(features.get("warnings") or [])

    return {
        "variant": parsed_variant,
        "classification_summary": {
            "alpha_missense": alpha,
            "clinvar": clinvar_records[:5],
            "gnomad": annotations.get("gnomad"),
            "overall_interpretation": _overall_interpretation(alpha, clinvar_records),
        },
        "wild_type": {
            "gene": parsed_variant.get("gene"),
            "protein": annotations.get("uniprot", {}).get("protein_name"),
            "uniprot": annotations.get("uniprot"),
            "sequence_length": annotations.get("uniprot", {}).get("length"),
            "description": _wild_type_description(annotations),
        },
        "unknown_variant": {
            "display_name": parsed_variant.get("display_name"),
            "mutation": parsed_variant.get("mutation"),
            "features": features,
            "description": _unknown_variant_description(parsed_variant, features),
        },
        "similar_variants": list(similar_variants),
        "structures": structures,
        "warnings": warnings,
    }


async def _claude_report(evidence: dict[str, Any]) -> str | None:
    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        return None
    try:
        import anthropic

        print(">>> Calling Claude API for report generation...", flush=True)
        client = anthropic.AsyncAnthropic(api_key=api_key)
        message = await client.messages.create(
            model=os.getenv("FOLDEX_CLAUDE_MODEL", "claude-sonnet-4-20250514"),
            max_tokens=1800,
            temperature=0,
            messages=[
                {
                    "role": "user",
                    "content": (
                        "Generate a concise variant research report in markdown using only the "
                        "structured evidence below. Do not invent symptoms, diagnoses, population "
                        "frequencies, pathogenicity, papers, or structures. If evidence is missing, "
                        "say it is missing. Include sections for classification, wild type, unknown "
                        "variant features, similar variants, likely behavior/expression based only on "
                        "similar known variants, 3D structures, and research-only disclaimer.\n\n"
                        f"{json.dumps({k: v for k, v in evidence.items() if k != 'structures'}, indent=2)}"
                    ),
                }
            ],
        )
        return "".join(block.text for block in message.content if hasattr(block, "text"))
    except Exception as e:
        print(f">>> Claude API failed: {e}", flush=True)
        return None


async def _groq_report(evidence: dict[str, Any]) -> str | None:
    api_key = os.getenv("GROQ_API_KEY")
    if not api_key:
        return None

    print(">>> Calling Groq API for report generation...", flush=True)
    url = os.getenv("FOLDEX_GROQ_URL", "https://api.groq.com/openai/v1/chat/completions")
    model = os.getenv("FOLDEX_GROQ_MODEL", "llama-3.3-70b-versatile")
    # Remove structures from the payload to stay under Groq size limits
    groq_evidence = {k: v for k, v in evidence.items() if k != "structures"}
    prompt = (
        "Generate a concise variant research report in markdown using only the structured "
        "evidence below. Do not invent symptoms, diagnoses, population frequencies, "
        "pathogenicity, papers, or structures. If evidence is missing, say it is missing. "
        "Include sections for classification, wild type, unknown variant features, similar "
        "variants, likely behavior/expression based only on similar known variants, 3D "
        "structures, and research-only disclaimer.\n\n"
        f"{json.dumps(groq_evidence, indent=2)}"
    )
    try:
        async with httpx.AsyncClient(timeout=120) as client:
            response = await client.post(
                url,
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": model,
                    "temperature": 0,
                    "messages": [{"role": "user", "content": prompt}],
                },
            )
            response.raise_for_status()
            data = response.json()
            choices = data.get("choices") or []
            if not choices:
                return None
            return choices[0].get("message", {}).get("content")
    except Exception as e:
        print(f">>> Groq API failed: {e}", flush=True)
        return None


def _fallback_markdown(evidence: dict[str, Any]) -> str:
    variant = evidence["variant"]
    summary = evidence["classification_summary"]
    unknown = evidence["unknown_variant"]
    wild_type = evidence["wild_type"]
    similar = evidence["similar_variants"]
    structures = evidence["structures"]
    warnings = evidence["warnings"]

    lines = [
        "# Variant Research Report",
        "",
        f"## Variant",
        f"- Input: {variant.get('input_text')}",
        f"- Parsed variant: {variant.get('display_name')}",
        f"- Gene: {variant.get('gene') or 'unknown'}",
        "",
        "## Benign / Pathogenic Evidence",
        f"- Overall interpretation: {summary['overall_interpretation']}",
        f"- AlphaMissense status: {summary.get('alpha_missense', {}).get('status')}",
        f"- ClinVar records found: {len(summary.get('clinvar') or [])}",
        f"- Population frequency: {summary.get('gnomad', {}).get('population_frequency')}",
        "",
        "## Wild Type",
        f"- Protein: {wild_type.get('protein') or 'unknown'}",
        f"- Sequence length: {wild_type.get('sequence_length') or 'unknown'}",
        f"- Description: {wild_type.get('description')}",
        "",
        "## Unknown Variant Features",
        f"- Description: {unknown.get('description')}",
        f"- Mutation features: {json.dumps(unknown.get('features', {}).get('mutation'), default=str)}",
        "",
        "## Similar Known Variants",
    ]
    if similar:
        for item in similar:
            lines.append(
                f"- {item.get('name')}: score {item.get('similarity_score')} "
                f"({'; '.join(item.get('similarity_reasons') or [])}). "
                f"{item.get('description')}"
            )
    else:
        lines.append("- No similar variants were found from the available evidence.")

    lines.extend(
        [
            "",
            "## Likely Behavior / Expression",
            _behavior_summary(similar),
            "",
            "## 3D Structures",
            f"- Wild type model available: {structures.get('wild_type', {}).get('viewer', {}).get('available')}",
            f"- Unknown variant model available: {structures.get('unknown_variant', {}).get('viewer', {}).get('available')}",
            f"- Similar variant models: {len(structures.get('similar_variants') or [])}",
            "",
            "## Warnings",
        ]
    )
    lines.extend([f"- {warning}" for warning in warnings] or ["- None"])
    lines.extend(
        [
            "",
            "Research support only. This is not medical advice, diagnosis, or treatment guidance.",
        ]
    )
    return "\n".join(lines)


def _overall_interpretation(alpha: dict[str, Any], clinvar_records: list[dict[str, Any]]) -> str:
    clinvar_labels = " ".join(
        (record.get("clinical_significance") or "").lower() for record in clinvar_records
    )
    if "pathogenic" in clinvar_labels and "benign" not in clinvar_labels:
        return "Known ClinVar evidence includes pathogenic assertions."
    if "benign" in clinvar_labels and "pathogenic" not in clinvar_labels:
        return "Known ClinVar evidence includes benign assertions."
    predictions = alpha.get("predictions") or []
    if predictions:
        return "AlphaMissense predictions are available and should be reviewed with ClinVar and frequency evidence."
    return "Insufficient evidence for benign/pathogenic classification."


def _wild_type_description(annotations: dict[str, Any]) -> str:
    uniprot = annotations.get("uniprot") or {}
    functions = uniprot.get("function") or []
    if functions:
        return functions[0]
    return "Wild-type function was not available from the current evidence."


def _unknown_variant_description(variant: dict[str, Any], features: dict[str, Any]) -> str:
    mutation = variant.get("mutation") or {}
    feature_mutation = features.get("mutation") or {}
    pieces = [
        f"Substitution {mutation.get('protein_hgvs')}" if mutation.get("protein_hgvs") else None,
        f"at protein position {mutation.get('protein_position')}"
        if mutation.get("protein_position")
        else None,
        f"changes amino-acid class: {feature_mutation.get('class_change')}"
        if feature_mutation.get("class_change") is not None
        else None,
        f"hydropathy delta {feature_mutation.get('hydropathy_delta')}"
        if feature_mutation.get("hydropathy_delta") is not None
        else None,
    ]
    return "; ".join(piece for piece in pieces if piece) or "Variant features are incomplete."


def _behavior_summary(similar_variants: list[dict[str, Any]]) -> str:
    known = [item for item in similar_variants if item.get("source") == "ClinVar"]
    if not known:
        return (
            "No known similar variants with clinical assertions were available, so behavior or "
            "expression should not be inferred beyond the measured features."
        )
    labels = ", ".join(item.get("clinical_significance") or "unknown" for item in known[:5])
    return f"Closest known variant assertions include: {labels}. Interpret cautiously with the ranked evidence."

File: server/reporter.ts
import { fetchWithTimeout } from './http'

type AnyDict = Record<string, unknown>

function asDict(v: unknown): AnyDict {
  return v && typeof v === 'object' ? (v as AnyDict) : {}
}

function asArray<T = unknown>(v: unknown): T[] {
  return Array.isArray(v) ? (v as T[]) : []
}

const PATIENT_SUMMARY_INSTRUCTIONS =
  'Rewrite the variant report below as a short plain-language explanation for the patient. ' +
  'Use only facts already present in the report — do not introduce new findings, statistics, ' +
  'citations, or recommendations. Audience is a non-scientist adult with no genetics background.\n\n' +
  'Requirements:\n' +
  '- 120-180 words, in markdown.\n' +
  '- Start with a one-sentence plain-language description of what the variant is.\n' +
  '- Use 3 short sections with these exact headings:\n' +
  '  **What we found**, **What this means**, **What to do next**.\n' +
  "- Translate jargon (e.g. 'missense', 'pathogenic', 'allele frequency'). No abbreviations " +
  'without an inline definition. No raw scores or p-values.\n' +
  '- Preserve uncertainty honestly: if the report says evidence is missing, conflicting, or ' +
  'uncertain, say so plainly. Do not reassure beyond the evidence.\n' +
  "- Under 'What to do next', the only recommendation is to discuss the report with their " +
  'clinician or genetic counselor. Do not suggest treatments, lifestyle changes, or tests.\n' +
  '- End with this exact line on its own:\n' +
  '  *This summary is for research support only and is not medical advice. Please review it ' +
  'with a qualified clinician or genetic counselor.*\n'

function patientSummaryPrompt(reportMarkdown: string, evidence: AnyDict): string {
  const variant = asDict(evidence.variant)
  const context = {
    display_name: variant.display_name ?? null,
    gene: variant.gene ?? null,
    overall_interpretation: asDict(evidence.classification_summary).overall_interpretation ?? null,
  }
  return (
    `${PATIENT_SUMMARY_INSTRUCTIONS}\n` +
    `Variant context (for grounding only, do not quote verbatim):\n` +
    `${JSON.stringify(context, null, 2)}\n\n` +
    `Full report markdown:\n---\n${reportMarkdown}\n---`
  )
}

function overallInterpretation(alpha: AnyDict, clinvarRecords: AnyDict[]): string {
  const labels = clinvarRecords
    .map((r) => ((r.clinical_significance as string) ?? '').toLowerCase())
    .join(' ')
  if (labels.includes('pathogenic') && !labels.includes('benign')) {
    return 'Known ClinVar evidence includes pathogenic assertions.'
  }
  if (labels.includes('benign') && !labels.includes('pathogenic')) {
    return 'Known ClinVar evidence includes benign assertions.'
  }
  const predictions = asArray(alpha.predictions)
  if (predictions.length) {
    return 'AlphaMissense predictions are available and should be reviewed with ClinVar and frequency evidence.'
  }
  return 'Insufficient evidence for benign/pathogenic classification.'
}

function wildTypeDescription(annotations: AnyDict): string {
  const uniprot = asDict(annotations.uniprot)
  const functions = asArray<string>(uniprot.function)
  if (functions.length) return functions[0]
  return 'Wild-type function was not available from the current evidence.'
}

function unknownVariantDescription(variant: AnyDict, features: AnyDict): string {
  const mutation = asDict(variant.mutation)
  const featureMutation = asDict(features.mutation)
  const pieces = [
    mutation.protein_hgvs ? `Substitution ${mutation.protein_hgvs}` : null,
    mutation.protein_position ? `at protein position ${mutation.protein_position}` : null,
    featureMutation.class_change !== undefined && featureMutation.class_change !== null
      ? `changes amino-acid class: ${featureMutation.class_change}`
      : null,
    featureMutation.hydropathy_delta !== undefined && featureMutation.hydropathy_delta !== null
      ? `hydropathy delta ${featureMutation.hydropathy_delta}`
      : null,
  ]
  return pieces.filter(Boolean).join('; ') || 'Variant features are incomplete.'
}

function structuredEvidence(
  variant: AnyDict,
  annotations: AnyDict,
  similarVariants: AnyDict[],
): AnyDict {
  const parsedVariant = asDict(annotations.variant) || variant
  const features = asDict(annotations.features)
  const alpha = asDict(annotations.alpha_missense)
  const clinvarRecords = asArray<AnyDict>(asDict(annotations.clinvar).records)
  const warnings: string[] = []
  warnings.push(...asArray<string>(annotations.warnings))
  warnings.push(...asArray<string>(features.warnings))

  return {
    variant: parsedVariant,
    classification_summary: {
      alpha_missense: alpha,
      clinvar: clinvarRecords.slice(0, 5),
      gnomad: annotations.gnomad ?? null,
      overall_interpretation: overallInterpretation(alpha, clinvarRecords),
    },
    wild_type: {
      gene: parsedVariant.gene ?? null,
      protein: asDict(annotations.uniprot).protein_name ?? null,
      uniprot: annotations.uniprot ?? null,
      sequence_length: asDict(annotations.uniprot).length ?? null,
      description: wildTypeDescription(annotations),
    },
    unknown_variant: {
      display_name: parsedVariant.display_name ?? null,
      mutation: parsedVariant.mutation ?? null,
      features,
      description: unknownVariantDescription(parsedVariant, features),
    },
    similar_variants: similarVariants.map((item) => ({
      name: item.name ?? null,
      source: item.source ?? null,
      similarity_score: item.similarity_score ?? null,
      similarity_reasons: item.similarity_reasons ?? null,
      clinical_significance: item.clinical_significance ?? null,
      description: item.description ?? null,
    })),
    warnings,
  }
}

async function claudeReport(evidence: AnyDict): Promise<string | null> {
  const apiKey = process.env.ANTHROPIC_API_KEY
  if (!apiKey) return null
  try {
    const Anthropic = (await import('@anthropic-ai/sdk')).default
    const client = new Anthropic({ apiKey })
    const evidenceForPrompt: AnyDict = {}
    for (const [k, v] of Object.entries(evidence)) {
      if (k !== 'structures') evidenceForPrompt[k] = v
    }
    const message = await client.messages.create({
      model: process.env.FOLDEX_CLAUDE_MODEL ?? 'claude-sonnet-4-20250514',
      max_tokens: 1800,
      temperature: 0,
      messages: [
        {
          role: 'user',
          content:
            'Generate a concise variant research report in markdown using only the ' +
            'structured evidence below. Do not invent symptoms, diagnoses, population ' +
            'frequencies, pathogenicity, papers, or structures. If evidence is missing, ' +
            'say it is missing. Include sections for classification, wild type, unknown ' +
            'variant features, similar variants, likely behavior/expression based only on ' +
            'similar known variants, 3D structures, and research-only disclaimer.\n\n' +
            JSON.stringify(evidenceForPrompt, null, 2),
        },
      ],
    })
    return message.content
      .map((block: { type: string; text?: string }) => (block.type === 'text' ? block.text ?? '' : ''))
      .join('')
  } catch (e) {
    console.error('Claude report failed:', e)
    return null
  }
}

async function claudePatientSummary(reportMarkdown: string, evidence: AnyDict): Promise<string | null> {
  const apiKey = process.env.ANTHROPIC_API_KEY
  if (!apiKey || !reportMarkdown) return null
  try {
    const Anthropic = (await import('@anthropic-ai/sdk')).default
    const client = new Anthropic({ apiKey })
    const message = await client.messages.create({
      model: process.env.FOLDEX_CLAUDE_MODEL ?? 'claude-sonnet-4-20250514',
      max_tokens: 600,
      temperature: 0,
      messages: [{ role: 'user', content: patientSummaryPrompt(reportMarkdown, evidence) }],
    })
    const text = message.content
      .map((block: { type: string; text?: string }) => (block.type === 'text' ? block.text ?? '' : ''))
      .join('')
      .trim()
    return text || null
  } catch (e) {
    console.error('Claude patient summary failed:', e)
    return null
  }
}

async function groqReport(evidence: AnyDict): Promise<string | null> {
  const apiKey = process.env.GROQ_API_KEY
  if (!apiKey) return null
  const url = process.env.FOLDEX_GROQ_URL ?? 'https://api.groq.com/openai/v1/chat/completions'
  const model = process.env.FOLDEX_GROQ_MODEL ?? 'llama-3.3-70b-versatile'
  const evidenceForPrompt: AnyDict = {}
  for (const [k, v] of Object.entries(evidence)) {
    if (k !== 'structures') evidenceForPrompt[k] = v
  }
  const prompt =
    'Generate a concise variant research report in markdown using only the structured ' +
    'evidence below. Do not invent symptoms, diagnoses, population frequencies, ' +
    'pathogenicity, papers, or structures. If evidence is missing, say it is missing. ' +
    'Include sections for classification, wild type, unknown variant features, similar ' +
    'variants, likely behavior/expression based only on similar known variants, 3D ' +
    'structures, and research-only disclaimer.\n\n' +
    JSON.stringify(evidenceForPrompt, null, 2)
  try {
    const response = await fetchWithTimeout(url, {
      method: 'POST',
      headers: { Authorization: `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model,
        messages: [{ role: 'user', content: prompt }],
        temperature: 0,
      }),
      timeoutMs: 50_000,
    })
    if (!response.ok) throw new Error(`HTTP ${response.status}`)
    const data = (await response.json()) as AnyDict
    const choices = asArray<AnyDict>(data.choices)
    if (!choices.length) return null
    return (asDict(choices[0].message).content as string) ?? null
  } catch (e) {
    console.error('Groq report failed:', e)
    return null
  }
}

async function groqPatientSummary(reportMarkdown: string, evidence: AnyDict): Promise<string | null> {
  const apiKey = process.env.GROQ_API_KEY
  if (!apiKey || !reportMarkdown) return null
  const url = process.env.FOLDEX_GROQ_URL ?? 'https://api.groq.com/openai/v1/chat/completions'
  const model = process.env.FOLDEX_GROQ_MODEL ?? 'llama-3.3-70b-versatile'
  try {
    const response = await fetchWithTimeout(url, {
      method: 'POST',
      headers: { Authorization: `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model,
        messages: [{ role: 'user', content: patientSummaryPrompt(reportMarkdown, evidence) }],
        temperature: 0,
      }),
      timeoutMs: 50_000,
    })
    if (!response.ok) throw new Error(`HTTP ${response.status}`)
    const data = (await response.json()) as AnyDict
    const choices = asArray<AnyDict>(data.choices)
    if (!choices.length) return null
    return ((asDict(choices[0].message).content as string) ?? '').trim() || null
  } catch (e) {
    console.error('Groq patient summary failed:', e)
    return null
  }
}

function fallbackPatientSummary(evidence: AnyDict): string {
  const variant = asDict(evidence.variant)
  const summary = asDict(evidence.classification_summary)
  const name = (variant.display_name as string) || (variant.gene as string) || 'this variant'
  const interpretation =
    (summary.overall_interpretation as string) ||
    'The available evidence is limited, so the effect of this change is uncertain.'
  return [
    '**What we found**\n',
    `Your test identified a change in your DNA called ${name}. A 'variant' simply means a spelling difference in a gene compared with the typical reference.\n`,
    '**What this means**\n',
    `${interpretation} In plain terms, scientists do not yet have enough information to say with confidence whether this change affects your health.\n`,
    '**What to do next**\n',
    'Please share this report with your clinician or a genetic counselor. They can put it in the context of your personal and family history and help you decide on next steps.\n',
    '*This summary is for research support only and is not medical advice. Please review it with a qualified clinician or genetic counselor.*',
  ].join('\n')
}

function behaviorSummary(similarVariants: AnyDict[]): string {
  const known = similarVariants.filter((item) => item.source === 'ClinVar')
  if (!known.length) {
    return 'No known similar variants with clinical assertions were available, so behavior or expression should not be inferred beyond the measured features.'
  }
  const labels = known
    .slice(0, 5)
    .map((item) => (item.clinical_significance as string) || 'unknown')
    .join(', ')
  return `Closest known variant assertions include: ${labels}. Interpret cautiously with the ranked evidence.`
}

function fallbackMarkdown(evidence: AnyDict, structures: AnyDict): string {
  const variant = asDict(evidence.variant)
  const summary = asDict(evidence.classification_summary)
  const unknown = asDict(evidence.unknown_variant)
  const wildType = asDict(evidence.wild_type)
  const similar = asArray<AnyDict>(evidence.similar_variants)
  const warnings = asArray<string>(evidence.warnings)

  const lines: string[] = [
    '# Variant Research Report',
    '',
    '## Variant',
    `- Input: ${variant.input_text}`,
    `- Parsed variant: ${variant.display_name}`,
    `- Gene: ${variant.gene || 'unknown'}`,
    '',
    '## Benign / Pathogenic Evidence',
    `- Overall interpretation: ${summary.overall_interpretation}`,
    `- AlphaMissense status: ${asDict(summary.alpha_missense).status ?? ''}`,
    `- ClinVar records found: ${asArray(summary.clinvar).length}`,
    `- Population frequency: ${asDict(summary.gnomad).population_frequency ?? ''}`,
    '',
    '## Wild Type',
    `- Protein: ${wildType.protein || 'unknown'}`,
    `- Sequence length: ${wildType.sequence_length || 'unknown'}`,
    `- Description: ${wildType.description}`,
    '',
    '## Unknown Variant Features',
    `- Description: ${unknown.description}`,
    `- Mutation features: ${JSON.stringify(asDict(unknown.features).mutation ?? {})}`,
    '',
    '## Similar Known Variants',
  ]

  if (similar.length) {
    for (const item of similar) {
      const reasons = asArray<string>(item.similarity_reasons).join('; ')
      lines.push(
        `- ${item.name}: score ${item.similarity_score} (${reasons}). ${item.description ?? ''}`,
      )
    }
  } else {
    lines.push('- No similar variants were found from the available evidence.')
  }

  lines.push(
    '',
    '## Likely Behavior / Expression',
    behaviorSummary(similar),
    '',
    '## 3D Structures',
    `- Wild type model available: ${asDict(asDict(structures.wild_type).viewer).available}`,
    `- Unknown variant model available: ${asDict(asDict(structures.unknown_variant).viewer).available}`,
    `- Similar variant models: ${asArray(structures.similar_variants).length}`,
    '',
    '## Warnings',
  )

  if (warnings.length) {
    for (const w of warnings) lines.push(`- ${w}`)
  } else {
    lines.push('- None')
  }

  lines.push('', 'Research support only. This is not medical advice, diagnosis, or treatment guidance.')
  return lines.join('\n')
}

export async function generateReport(
  variant: AnyDict,
  annotations: AnyDict,
  similarVariants: AnyDict[],
  structures: AnyDict,
): Promise<AnyDict> {
  const evidence = structuredEvidence(variant, annotations, similarVariants)
  const aiMarkdown = (await claudeReport(evidence)) ?? (await groqReport(evidence))
  const markdown = aiMarkdown ?? fallbackMarkdown(evidence, structures)

  const patientSummary =
    (await claudePatientSummary(markdown, evidence)) ??
    (await groqPatientSummary(markdown, evidence)) ??
    fallbackPatientSummary(evidence)

  return {
    markdown,
    patient_summary: patientSummary,
    json: {
      status: 'complete',
      variant: evidence.variant,
      classification_summary: evidence.classification_summary,
      wild_type: evidence.wild_type,
      unknown_variant: evidence.unknown_variant,
      similar_variants: evidence.similar_variants,
      structures,
      warnings: evidence.warnings,
      disclaimer: 'Research support only. This is not medical advice, diagnosis, or treatment guidance.',
    },
  }
}
